game-app/game-core/src/main/java/games/strategy/engine/data/Rule.java

```diff
@@ -20,17 +20,7 @@ public interface Rule {
 
   IntegerMap<NamedAttachable> getResults();
 
-  IntegerMap<Resource> getInternalCosts();
-
-  /**
-   * Add resource cost to rule.
-   *
-   * @param resource of the cost
-   * @param quantity of the {@code resource}
-   */
-  default void addCost(final Resource resource, final int quantity) {
-    getInternalCosts().put(resource, quantity);
-  }
+  void addCost(final Resource resource, final int quantity);
```

Comment:
We have one less method on the interface and we no longer require implementations to expose mutable internal data structures